File: static/js/pesquisar_cliente.js
document.addEventListener("DOMContentLoaded", function () {
    let selectedRow = null; // Variable to store the currently selected row

    // Function to handle row selection: ensure only one row can be selected at a time
    function handleRowClick(row) {
        // If a row is already selected, remove the 'selected' class from it
        if (selectedRow) {
            selectedRow.classList.remove('selected');
        }

        // Add the 'selected' class to the clicked row
        row.classList.add('selected');
        selectedRow = row;  // Update the selected row
    }

    // Add click event listeners to each row in the table
    document.querySelectorAll('.results-table .client-row').forEach(function (row) {
        row.addEventListener('click', function () {
            handleRowClick(row);
        });
    });

    // "Edit" button functionality: navigate to edit page with the client ID
    document.getElementById('editButton').addEventListener('click', function () {
        if (selectedRow) {
            const clientId = selectedRow.cells[0].textContent; // Assuming the client ID is in the first column
            window.location.href = `cliente_cadastrar/${clientId}`;  // Redirect to the edit page for the selected client
        } else {
            alert('Nenhum cliente selecionado para editar.');
        }
    });

    // "Delete" button functionality: confirm deletion and delete the client
    document.getElementById('deleteButton').addEventListener('click', function () {
        if (selectedRow) {
            const clientId = selectedRow.cells[0].textContent; // Assuming the client ID is in the first column
            const confirmDelete = confirm(`Tem certeza que deseja deletar o cliente com código ${clientId}?`);
            if (confirmDelete) {
                // Proceed with deleting the selected client
                const formData = new FormData();
                formData.append('id', clientId);

                fetch('/delete_cliente', {
                    method: 'POST',
                    body: formData
                })
                    .then(response => {
                        if (response.ok) {
                            // Remove the selected row from the table
                            selectedRow.remove();
                            selectedRow = null;  // Deselect after deleting
                        } else {
                            alert('Erro ao deletar o cliente.');
                        }
                    })
                    .catch(error => console.error('Error during fetch:', error));
            }
        } else {
            alert('Nenhum cliente selecionado para deletar.');
        }
    });

    // Search form validation logic
    function validateForm(event) {
        event.preventDefault(); // Prevent the form from submitting right away
        const formValues = {
            id: document.getElementById('id')?.value ?? '',
            nome: document.getElementById('nome')?.value ?? '',
            telefone: document.getElementById('telefone')?.value ?? '',
            endereco: document.getElementById('endereco')?.value ?? ''
        };
        const errorDiv = document.getElementById('error-message-container');
        errorDiv.innerHTML = '';
        errorDiv.style.color = 'red'; // Set error message color
        const validationRules = {
            id: {
                regex: /^[1-9][0-9]*$/, // Allows positive numbers greater than or equal to 1
                errorMessage: 'Código deve ser um número positivo maior ou igual a 1.'
            },
            nome: {
                regex: /^[A-Za-zÀ-ÿ\s]*$/, // Allows letters and spaces or empty
                errorMessage: 'O campo Nome é obrigatório e deve conter apenas letras.'
            },
            telefone: {
                regex: /^\(\d{2}\)\d{8,9}$/, // Matches phone format like (99)999999999 or (99)99999999
                errorMessage: 'O telefone é obrigatório e deve estar no formato (99)999999999 ou (99)99999999.'
            },
            endereco: {
                regex: /^[A-Za-zÀ-ÿ0-9\s,.-]*$/, // Allows letters, numbers, spaces, commas, periods, and hyphens
                errorMessage: 'Endereço contém caracteres inválidos.'
            }
        };
        for (const field in formValues) {
            const value = formValues[field];
            const rule = validationRules[field];
            if (value && rule && !rule.regex.test(value)) {
                errorDiv.innerHTML = rule.errorMessage;
                return false;
            }
            if (field === 'id' && value !== '' && !/^[1-9]\d*$/.test(value)) {
                errorDiv.innerHTML = 'O Código deve ser um número positivo maior ou igual a 1.';
                return false;
            }
        }
        return true;
    }

    // Handle the search button click event
    document.getElementById('pesquisarButton').addEventListener('click', function (event) {
        const isValid = validateForm(event); // Call the validation function
        if (isValid) {
            const formData = new FormData(document.getElementById('searchForm'));
            fetch('/pesquisar_cliente', {
                method: 'POST',
                body: formData
            })
                .then(response => {
                    if (!response.ok) {
                        return response.text().then(text => {
                            throw new Error(text || 'Erro ocorreu ao buscar dados');
                        });
                    }
                    return response.json(); // Parse the JSON response
                })
                .then(data => {
                    // Clear previous results and error messages
                    const clientTableContainer = document.getElementById('clientTableContainer');
                    clientTableContainer.innerHTML = ''; // Clear previous results
                    clientTableContainer.style.display = ''; // Make sure the container is visible
                    document.querySelector('.error-message')?.remove(); // Remove any previous error message

                    if (data.clientes && data.clientes.length > 0) {
                        let tableHTML = `
                            <table class="results-table">
                                <thead>
                                    <tr>
                                        <th>Código</th>
                                        <th>Nome</th>
                                        <th>Endereço</th>
                                        <th>Número</th>
                                        <th>Bairro</th>
                                        <th>Complemento</th>
                                        <th>Observação</th>
                                        <th>Telefones</th>
                                    </tr>
                                </thead>
                                <tbody>
                        `;

                        data.clientes.forEach(cliente => {
                            tableHTML += `
                                <tr class="client-row">
                                    <td>${cliente.id}</td>
                                    <td>${cliente.nome}</td>
                                    <td>${cliente.endereco}</td>
                                    <td>${cliente.numero}</td>
                                    <td>${cliente.bairro}</td>
                                    <td>${cliente.complemento}</td>
                                    <td>${cliente.observacao}</td>
                                    <td>
                                        <div class="telefone-container">
                                            <span>${cliente.telefones.join(', ')}</span>
                                        </div>
                                    </td>
                                </tr>
                            `;
                        });
                        tableHTML += `</tbody></table>`;
                        clientTableContainer.innerHTML = tableHTML;

                        // Reapply the row selection event listeners after populating the table
                        document.querySelectorAll('.results-table .client-row').forEach(function (row) {
                            row.addEventListener('click', function () {
                                handleRowClick(row);
                            });
                        });
                    } else {
                        clientTableContainer.innerHTML = '<p>Nenhum cliente encontrado.</p>';
                    }
                })
                .catch(error => {
                    const existingErrorDiv = document.querySelector('.error-message');
                    if (existingErrorDiv) {
                        existingErrorDiv.remove(); // Remove the old error message
                    }

                    const errorDiv = document.createElement('div');
                    errorDiv.className = 'error-message';
                    errorDiv.style.color = 'red';

                    errorDiv.textContent = 'Erro ao processar a resposta do servidor';
                    document.getElementById('error-message-container').appendChild(errorDiv);
                    console.error('Error during fetch:', error);
                });
        }
    });
});

function formatCellphone(input) {
    // Get the input value without any non-digit characters
    let value = input.value.replace(/\D/g, '');

    // Check the length of the value and format accordingly
    if (value.length <= 2) {
        // Format as (XX)
        input.value = '(' + value;
    } else if (value.length <= 6) {
        // Format as (XX)XXXXX
        input.value = '(' + value.substring(0, 2) + ')' + value.substring(2);
    } else {
        // Format as (XX)XXXXXXXXX
        input.value = '(' + value.substring(0, 2) + ')' + value.substring(2, 7) + value.substring(7, 11);
    }
}

File: static/py/pesquisar_cliente.py
from flask import Flask, render_template, request, Blueprint, flash, redirect, url_for, jsonify
import psycopg2
from static.py.config.db import get_db_connection
from static.py.login_required import login_required
import re

pesquisar_cliente_bp = Blueprint('pesquisar_cliente_bp', __name__)

@pesquisar_cliente_bp.route('/pesquisar_cliente', methods=['GET', 'POST'])
@login_required
def pesquisar_cliente():
    conn = get_db_connection()
    cur = conn.cursor()
    
    if request.method == 'POST':
        codigo = request.form.get('id', '').strip()
        nome = request.form.get('nome', '').strip()
        telefone = request.form.get('telefone', '').strip()
        endereco = request.form.get('endereco', '').strip()

        # Check if at least one filter is provided
        if not (codigo or nome or telefone or endereco):
            return jsonify({'error': "Por favor, insira pelo menos um filtro para a pesquisa."}), 400

        # Construct the base query
        query = """
            SELECT c.id, c.nome, c.endereco, c.numero, c.bairro, c.complemento, c.observacao,
                array_agg(t.telefone) AS telefones
            FROM cliente c
            LEFT JOIN telefone t ON c.id = t.cliente_id
        """
        
        # List to hold parameters
        parameters = []

        # Add conditions based on the provided filters
        query_conditions = ["WHERE 1=1"]  # Base condition to make it easier to add more
        
        if codigo:
            query_conditions.append(" AND c.id = %s")
            parameters.append(int(codigo))
        if nome:
            query_conditions.append(" AND c.nome ILIKE %s")
            parameters.append(f'%{nome}%')
        if telefone:
            query_conditions.append(" AND (t.telefone ILIKE %s OR c.id IN (SELECT cliente_id FROM telefone WHERE telefone ILIKE %s))")
            parameters.append(f'%{telefone}%')
            parameters.append(f'%{telefone}%')  # To match the condition for the subquery
        if endereco:
            query_conditions.append(" AND c.endereco ILIKE %s")
            parameters.append(f'%{endereco}%')
        
        query += " ".join(query_conditions)

        # Add GROUP BY clause to aggregate phone numbers
        query += """
            GROUP BY c.id
        """

        try:
            # Execute the query with dynamic parameters
            cur.execute(query, parameters)
            clientes_data = cur.fetchall()

            # Prepare the results
            clientes = []
            for cliente in clientes_data:
                cliente_dict = {
                    'id': cliente[0],
                    'nome': cliente[1],
                    'endereco': cliente[2],
                    'numero': cliente[3],
                    'bairro': cliente[4],
                    'complemento': cliente[5],
                    'observacao': cliente[6],
                    'telefones': cliente[7] if cliente[7] else []  # Ensure telefones is not None
                }
                clientes.append(cliente_dict)

            cur.close()
            conn.close()
            # Return the result
            if clientes:
                clientes.sort(key=lambda row: row['id'])
                return jsonify({'clientes': clientes}), 200
            else:
                return jsonify({'error': "Nenhum cliente encontrado"}), 404
        except Exception as e:
            return jsonify({'error': str(e)}), 500
    return render_template('pesquisar_cliente.html')

@pesquisar_cliente_bp.route('/delete_cliente', methods=['POST', 'GET'])
@login_required
def delete_cliente():
    cliente_id = request.form.get('id', '')
    print(cliente_id)
    if cliente_id:
        try:
            conn = get_db_connection()
            cur = conn.cursor()
            cur.execute('DELETE FROM cliente WHERE id = %s', (cliente_id,))
            conn.commit()
            cur.close()
            conn.close()
            flash('Cliente deletado com sucesso', 'success')
        except Exception as e:
            print(f"Error occurred: {e}")
            flash('Erro ao deletar cliente', 'error')
    else:
        flash('Nenhum ID de cliente fornecido', 'warning')

    return redirect(url_for('pesquisar_cliente_bp.pesquisar_cliente'))
